File: test_sort.py
import random
import nose
import sort
import quicksort as qs

LENGTH = 10

array = []

def setup():
    global array
    array = list(range(LENGTH))

def is_sorted(a):
    for i in range(len(a)-1):
        if a[i] > a[i+1]:
            return False
    return True
    
@nose.with_setup(setup)    
def test_insertsort():
    for i in range(10):
        random.shuffle(array)
        sorted = sort.insertsort(array)
        #print(sorted)
        assert is_sorted(sorted)


@nose.with_setup(setup)
def test_selectionsort():
    for i in range(10):
        random.shuffle(array)
        sorted = sort.selectionsort(array)
        assert is_sorted(sorted)

@nose.with_setup(setup)
def test_mergesort():
    random.shuffle(array)
    sorted = sort.mergesort(array)
    assert is_sorted(sorted)

@nose.with_setup(setup)
def test_quicksort():
    random.shuffle(array)
    sorted = qs.quicksort(array)
    assert is_sorted(sorted)

File: test_sort.cpp
#include <random>
#include "gtest/gtest.h"
#include "sort.h"
#include "utility.h"

const int LENGTH = 10000;
const int MIN = -1000;
const int MAX = 1000;

TEST(SortTest, InsertSort) {
    for(int i=0; i<10; ++i){
	int array[LENGTH] = {};
	GenerateUnsortedArray(array, LENGTH);
	InsertSort(array, LENGTH);
	EXPECT_TRUE(IsOrdered(array, LENGTH));
    }
}

TEST(SortTest, SelectionSort) {
    for(int i=0; i<10; ++i){
	int array[LENGTH] = {};
	GenerateUnsortedArray(array, LENGTH);
	SelectionSort(array, LENGTH);
	EXPECT_TRUE(IsOrdered(array, LENGTH));
    }
} 

TEST(SortTest, MergeSort) {
    for(int i=0; i<10; ++i){
	int array[LENGTH] = {};
	GenerateUnsortedArray(array, LENGTH);
	MergeSort(array, LENGTH);
	EXPECT_TRUE(IsOrdered(array, LENGTH));
    }
}

TEST(SortTest, RecursiveInsertSort) {
    for(int i=0; i<10; ++i){
	int array[LENGTH] = {};
	GenerateUnsortedArray(array, LENGTH);
	RecursiveInsertSort(array, LENGTH);
	EXPECT_TRUE(IsOrdered(array, LENGTH));
    }
}

TEST(SortTest, QuickSort) {
    for(int i=0; i<10; ++i){
	int array[LENGTH] = {};
	GenerateUnsortedArray(array, LENGTH);
	QuickSort(array, 0, LENGTH-1);
	EXPECT_TRUE(IsOrdered(array, LENGTH));
    }
}

int main(int argc, char **argv)
{
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
